File: RDE/data/SDToolbox/MATLAB/Demo/demo_shock_point.m
% Shock and Detonation Toolbox Demo Program
% 
% This is a demostration of how to compute frozen and equilibrium postshock
% conditions for a single shock Mach number
% 
% ################################################################################
% Theory, numerical methods and applications are described in the following report:
% 
%     Numerical Solution Methods for Shock and Detonation Jump Conditions, S.
%     Browne, J. Ziegler, and J. E. Shepherd, GALCIT Report FM2006.006 - R3,
%     California Institute of Technology Revised September, 2018.
% 
% Please cite this report and the website if you use these routines. 
% 
% Please refer to LICENCE.txt or the above report for copyright and disclaimers.
% 
% http://shepherd.caltech.edu/EDL/PublicResources/sdt/
% 
% ################################################################################ 
% Updated September 2018
% Tested with: 
%     MATLAB 2017b and 2018a, Cantera 2.3 and 2.4
% Under these operating systems:
%     Windows 8.1, Windows 10, Linux (Debian 9)
%%
close all;clear;clc;
disp('shock_point')
transport = true;
mech = 'airNASA9noions.cti';
gas1 = Solution(mech);
gas2 = Solution(mech);
gas3 = Solution(mech);
%% set initial state of gas.
P1 = 770.;
T1 = 728;
q = 'O2:1 N2:3.76';
set(gas1,'Temperature',T1,'Pressure',P1,'MoleFractions',q);
rho1 = density(gas1);
af1 = soundspeed_fr(gas1);
M1 = 7.16;   % shock Mach number
speed = M1*af1;
gamma1_fr = af1^2*rho1/P1;
if (transport)
    mu1 = viscosity(gas1);
    nu1 = mu1/rho1;
    kcond1 = thermalConductivity(gas1);
    kdiff1 = kcond1/(rho1*cp_mass(gas1));
    Pr = mu1*cp_mass(gas1)/kcond1;
end

%%  Print out initial state
disp(['Shock computation for ',mech,' with composition ',q]);
disp(['Initial State']);
disp(['   Shock speed ',num2str(speed),' (m/s)']);
disp(['   Frozen sound Speed ',num2str(af1),' (m/s)']);
disp(['   Shock Mach number ',num2str(M1)]);
disp(['   Pressure ',num2str(P1),' (Pa)']);
disp(['   Temperature ',num2str(T1),' (K)']);
disp(['   Density ',num2str(rho1),' (kg/m3)']);
disp(['   gamma2 (based on frozen sound speed) ',num2str(gamma1_fr),' (m/s)']);
disp(['   Specific heat at constant pressure ',num2str(cp_mass(gas1)),' (J/kg K)']);
if (transport)
    disp(['   viscosity ',num2str(mu1),' (kg/m s)']);
    disp(['   viscosity (kinematic)',num2str(nu1),' (m2/s)']);
    disp(['   thermal conductivity ',num2str(kcond1),' (W/m K)']);
    disp(['   thermal diffusivity ',num2str(kdiff1),' (m2/s)']);
    disp(['   Prandtl number ',num2str(Pr)]);
end

%% Postshock (Frozen)
[gas2] = PostShock_fr(speed,P1, T1, q, mech);
af2 = soundspeed_fr(gas2);
ae2 = soundspeed_eq(gas2);
P2 = pressure(gas2);
T2 = temperature(gas2);
S2 = entropy_mass(gas2);
rho2 = density(gas2);
gamma2_fr = af2*af2*rho2/P2;
gamma2_eq = ae2*ae2*rho2/P2;
w2 = rho1*speed/rho2;
u2 = speed - w2;
if (transport)
    mu = viscosity(gas2);
    nu = mu/rho2;
    kcond = thermalConductivity(gas2);
    kdiff = kcond/(rho2*cp_mass(gas2));
    Pr = mu*cp_mass(gas2)/kcond;
end

%%  Print frozen shock state
disp(['Frozen shock state'])
disp(['   Pressure ',num2str(P2),' (Pa)']);
disp(['   Temperature ',num2str(T2),' (K)']);
disp(['   Density ',num2str(rho2),' (kg/m3)']);
disp(['   Specific heat at constant pressure ',num2str(cp_mass(gas2)),' (J/kg K)']);
disp(['   Expansion  ratio ',num2str(rho1/rho2),' (kg/m3)']);
disp(['   Entropy ',num2str(S2),' (J/kg-K)']);
disp(['   Sound speed (frozen) ',num2str(af2),' (m/s)']);
disp(['   Sound speed (equilibrium) ',num2str(ae2),' (m/s)']);
disp(['   gamma2 (based on frozen sound speed) ',num2str(gamma2_fr),' (m/s)']);
disp(['   gamma2 (based on equilibrium sound speed) ',num2str(gamma2_eq),' (m/s)']);
if (transport)
    disp(['   viscosity ',num2str(mu),' (kg/m s)']);
    disp(['   viscosity (kinematic)',num2str(nu),' (m2/s)']);
    disp(['   thermal conductivity ',num2str(kcond),' (W/m K)']);
    disp(['   thermal diffusivity ',num2str(kdiff),' (m2/s)']);
    disp(['   Prandtl number ',num2str(Pr)]);
end

%%  Postshock (Equilibrium)
[gas3] = PostShock_eq(speed,P1, T1, q, mech);
af3 = soundspeed_fr(gas3);
ae3 = soundspeed_eq(gas3);
P3 = pressure(gas3);
T3 = temperature(gas3);
rho3 = density(gas3);
S3 = entropy_mass(gas3);
gamma3_fr = af3*af3*rho3/P3;
gamma3_eq =  ae3*ae3*rho3/P3;
w3 = rho1*speed/rho3;
u3 = speed - w3;
if (transport)
    mu = viscosity(gas3);
    nu = mu/rho3;
    kcond = thermalConductivity(gas3);
    kdiff = kcond/(rho3*cp_mass(gas3));
    Pr = mu*cp_mass(gas3)/kcond;
end

%%  Print equilibrium shock state
disp(['Equilibrium shock state'])
disp(['   Pressure ',num2str(P3),' (Pa)']);
disp(['   Temperature ',num2str(T3),' (K)']);
disp(['   Density ',num2str(rho3),' (kg/m3)']);
disp(['   Specific heat at constant pressure ',num2str(cp_mass(gas3)),' (J/kg K)']);
disp(['   Expansion  ratio ',num2str(rho1/rho3),' (kg/m3)']);
disp(['   Entropy ',num2str(S3),' (J/kg-K)']);
disp(['   Sound speed (frozen) ',num2str(af3),' (m/s)']);
disp(['   Sound speed (equilibrium) ',num2str(ae3),' (m/s)']);
disp(['   gamma2 (based on frozen sound speed) ',num2str(gamma3_fr),' (m/s)']);
disp(['   gamma2 (based on equilibrium sound speed) ',num2str(gamma3_eq),' (m/s)']);
if (transport)
    disp(['   viscosity ',num2str(mu),' (kg/m s)']);
    disp(['   viscosity (kinematic)',num2str(nu),' (m2/s)']);
    disp(['   thermal conductivity ',num2str(kcond),' (W/m K)']);
    disp(['   thermal diffusivity ',num2str(kdiff),' (m2/s)']);
    disp(['   Prandtl number ',num2str(Pr)]);
end

File: RDE/data/SDToolbox/Python3/demo/demo_shock_point.py
"""
Shock and Detonation Toolbox Demo Program

This is a demostration of how to compute frozen and equilibrium postshock
conditions for a single shock Mach number

################################################################################
Theory, numerical methods and applications are described in the following report:

SDToolbox - Numerical Tools for Shock and Detonation Wave Modeling,
Explosion Dynamics Laboratory, Contributors: S. Browne, J. Ziegler,
N. Bitter, B. Schmidt, J. Lawson and J. E. Shepherd, GALCIT
Technical Report FM2018.001 Revised January 2021.

Please cite this report and the website if you use these routines. 

Please refer to LICENCE.txt or the above report for copyright and disclaimers.

http://shepherd.caltech.edu/EDL/PublicResources/sdt/

################################################################################ 
Updated September 2018
Tested with: 
    Python 3.5 and 3.6, Cantera 2.3 and 2.4
Under these operating systems:
    Windows 8.1, Windows 10, Linux (Debian 9)
"""
import cantera as ct
from sdtoolbox.thermo import soundspeed_eq,soundspeed_fr
from sdtoolbox.postshock import PostShock_eq,PostShock_fr

transport = True
mech = 'airNASA9noions.cti'
gas1 = ct.Solution(mech)
gas2 = ct.Solution(mech)
gas3 = ct.Solution(mech)

## set initial state of gas.
P1 = 770
T1 = 728
q = 'O2:1 N2:3.76'
gas1.TPX = T1,P1,q
rho1 = gas1.density
af1 = soundspeed_fr(gas1)
M1 = 7.16 # shock Mach number
speed = M1*af1
gamma1_fr = af1**2*rho1/P1
if transport:
    mu1 = gas1.viscosity
    nu1 = mu1/rho1
    kcond1 = gas1.thermal_conductivity
    kdiff1 = kcond1/(rho1*gas1.cp_mass)
    Pr = mu1*gas1.cp_mass/kcond1

## Print out initial state
print('Shock computation for '+mech+' with composition '+q)
print('Initial State')
print('   Shock speed '+str(speed)+' (m/s)')
print('   Frozen sound Speed '+str(af1)+' (m/s)')
print('   Shock Mach number '+str(M1))
print('   Pressure '+str(P1)+' (Pa)')
print('   Temperature '+str(T1)+' (K)')
print('   Density '+str(rho1)+' (kg/m3)')
print('   gamma2 (based on frozen sound speed) '+str(gamma1_fr)+' (m/s)')
print('   Specific heat at constant pressure '+str(gas1.cp_mass)+' (J/kg K)')
if transport:
    print('   viscosity '+str(mu1)+' (kg/m s)')
    print('   viscosity (kinematic)'+str(nu1)+' (m2/s)')
    print('   thermal conductivity '+str(kcond1)+' (W/m K)')
    print('   thermal diffusivity '+str(kdiff1)+' (m2/s)')
    print('   Prandtl number '+str(Pr))

## Postshock (Frozen)
gas2 = PostShock_fr(speed,P1,T1,q,mech)
af2 = soundspeed_fr(gas2)
ae2 = soundspeed_eq(gas2)
P2 = gas2.P
T2 = gas2.T
S2 = gas2.entropy_mass
rho2 = gas2.density
gamma2_fr = af2**2*rho2/P2
gamma2_eq = ae2**2*rho2/P2
w2 = rho1*speed/rho2
u2 = speed - w2
if transport:
    mu = gas2.viscosity
    nu = mu/rho2
    kcond = gas2.thermal_conductivity
    kdiff = kcond/(rho2*gas2.cp_mass)
    Pr = mu*gas2.cp_mass/kcond

## Print frozen shock state
print('Frozen shock state')
print('   Pressure '+str(P2)+' (Pa)')
print('   Temperature '+str(T2)+' (K)')
print('   Density '+str(rho2)+' (kg/m3)')
print('   Specific heat at constant pressure '+str(gas2.cp_mass)+' (J/kg K)')
print('   Expansion  ratio '+str(rho1/rho2)+' (kg/m3)')
print('   Entropy '+str(S2)+' (J/kg-K)')
print('   Sound speed (frozen) '+str(af2)+' (m/s)')
print('   Sound speed (equilibrium) '+str(ae2)+' (m/s)')
print('   gamma2 (based on frozen sound speed) '+str(gamma2_fr)+' (m/s)')
print('   gamma2 (based on equilibrium sound speed) '+str(gamma2_eq)+' (m/s)')
if transport:
    print('   viscosity '+str(mu)+' (kg/m s)')
    print('   viscosity (kinematic)'+str(nu)+' (m2/s)')
    print('   thermal conductivity '+str(kcond)+' (W/m K)')
    print('   thermal diffusivity '+str(kdiff)+' (m2/s)')
    print('   Prandtl number '+str(Pr))

##  Postshock (Equilibrium)
gas3 = PostShock_eq(speed,P1,T1,q,mech)
af3 = soundspeed_fr(gas3)
ae3 = soundspeed_eq(gas3)
P3 = gas3.P
T3 = gas3.T
rho3 = gas3.density
S3 = gas3.entropy_mass
gamma3_fr = af3**2*rho3/P3
gamma3_eq = ae3**2*rho3/P3
w3 = rho1*speed/rho3
u3 = speed - w3
if transport:
    mu = gas3.viscosity
    nu = mu/rho3
    kcond = gas3.thermal_conductivity
    kdiff = kcond/(rho3*gas3.cp_mass)
    Pr = mu*gas3.cp_mass/kcond

##  Print equilibrium shock state
print('Equilibrium shock state')
print('   Pressure '+str(P3)+' (Pa)')
print('   Temperature '+str(T3)+' (K)')
print('   Density '+str(rho3)+' (kg/m3)')
print('   Specific heat at constant pressure '+str(gas3.cp_mass)+' (J/kg K)')
print('   Expansion  ratio '+str(rho1/rho3)+' (kg/m3)')
print('   Entropy '+str(S3)+' (J/kg-K)')
print('   Sound speed (frozen) '+str(af3)+' (m/s)')
print('   Sound speed (equilibrium) '+str(ae3)+' (m/s)')
print('   gamma2 (based on frozen sound speed) '+str(gamma3_fr)+' (m/s)')
print('   gamma2 (based on equilibrium sound speed) '+str(gamma3_eq)+' (m/s)')
if transport:
    print('   viscosity '+str(mu)+' (kg/m s)')
    print('   viscosity (kinematic)'+str(nu)+' (m2/s)')
    print('   thermal conductivity '+str(kcond)+' (W/m K)')
    print('   thermal diffusivity '+str(kdiff)+' (m2/s)')
    print('   Prandtl number '+str(Pr))
